Default to NODE_WITNESS in nLocalServices

dhruv opened the pull request:
Builds on #21009 and makes progress on remaining items in #17862

Removing `RewindBlockIndex()` in #21009 allows the following:

- removal of tests using `segwitheight=-1` in `p2p_segwit.py`.
- move `test_upgrade_after_activation()` out of `p2p_segwit.py` reducing runtime
- in turn, that allows us to drop support for `-segwitheight=-1`, which is only supported for that test.
- that allows us to always set `NODE_WITNESS` in our local services. The only reason we don't do that is to support `-segwitheight=-1`.
- that in turn allows us to drop all of the `GetLocalServices() & NODE_WITNESS` checks inside `net_processing.cpp`, since our local services would always include `NODE_WITNESS`

dhruv committed:
[test] remove or move tests using `-segwitheight=-1`

dhruv committed:
[p2p] remove unused segwitheight=-1 option

This also lets us default to NODE_WITNESS in nLocalServices

dhruv committed:
[p2p] remove redundant NODE_WITNESS checks

nLocalServices defaults to NODE_WITNESS and these checks are obsolete

jnewbery commented:
This needs rebase now that #21009 is merged. I did my own rebase here: https://github.com/jnewbery/bitcoin/tree/pr21090.1 since I need it for #20799. Feel free to take that if it's helpful.

I think it might make sense to remove the `test_upgrade_after_activation()` subtest from p2p_segwit.py and place it in its own file. It's not anything to do with p2p and doesn't have any dependencies on the rest of that test.

dhruv commented:
Rebased against master, ready for review.

@jnewbery it does make a ton of sense to extract `test_upgrade_after_activation()` into another file. Once `-reindex` has the expected effect, the rest of the test merely verifies that the node syncs on to the strongest chain, and that is in covered in other tests. I've gone ahead and done that refactor. This also simplified `p2p_segwit.py` since it now only needs two nodes for the rest of the test cases.

MarcoFalke commented on src/chainparams.cpp:
```diff
@@ -490,11 +490,8 @@ void CRegTestParams::UpdateActivationParametersFromArgs(const ArgsManager& args)
         int64_t height = args.GetArg("-segwitheight", consensus.SegwitHeight);
-        if (height < -1 || height >= std::numeric_limits<int>::max()) {
+        if (height < 0 || height >= std::numeric_limits<int>::max()) {
             throw std::runtime_error(strprintf("Activation height %ld for segwit is out of valid range. Use -1 to disable segwit.", height));
```
```suggestion
            throw std::runtime_error(strprintf("Activation height %ld for segwit is out of valid range.", height));
```

dhruv replied:
Done.

dhruv commented on src/net_processing.cpp:
```diff
@@ -2586,7 +2585,7 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         uint64_t nCMPCTBLOCKVersion = 0;
         vRecv >> fAnnounceUsingCMPCTBLOCK >> nCMPCTBLOCKVersion;
-        if (nCMPCTBLOCKVersion == 1 || ((pfrom.GetLocalServices() & NODE_WITNESS) && nCMPCTBLOCKVersion == 2)) {
+        if (nCMPCTBLOCKVersion == 1 || nCMPCTBLOCKVersion == 2) {
```
It [think](https://developer.bitcoin.org/reference/p2p_networking.html#sendcmpct) the only allowed values are 1 or 2. Can this conditional be removed altogether? Or perhaps replaced with `if (0 < nCMPCTBLOCKVersion < 3)` ?

jnewbery replied:
Yes, the only permitted values for nCMPCTBLOCKVersion are `1` and `2`, which is why this conditional is here. We want to ignore any `sendcmpct` message received where nCMPCTBLOCKVersion is not 1 or 2.

`if (0 < nCMPCTBLOCKVersion < 3)` does not do what you I think you expect it to do.

dhruv replied:
Ok, I've leave the conditional in there then! :)

dhruv reviewed (commented):
Rebased with master. Weirdly DrahtBot didn't mention it was needed (but the CI did).

Ready for further review with one question below.

jnewbery commented on test/functional/p2p_segwit.py:
```diff
@@ -224,7 +222,6 @@ def skip_test_if_missing_module(self):
     def setup_network(self):
```
This method override can be removed entirely, since it's doing the same as the default `setup_network()` in test_framework.py

dhruv replied:
Done

jnewbery commented on test/functional/feature_presegwit_node_upgrade.py:
```diff
@@ -0,0 +1,58 @@
+        self.setup_clean_chain = True
+        self.num_nodes = 1
+
+    def setup_network(self):
+        self.setup_nodes()
+
```
Not needed. This is default behaviour.

dhruv replied:
Done

jnewbery commented on test/functional/feature_presegwit_node_upgrade.py:
```diff
@@ -0,0 +1,58 @@
+    def setup_network(self):
+        self.setup_nodes()
+
+    def run_test(self):
+        self.test_upgrade_after_activation()
+
+    def test_upgrade_after_activation(self):
+        """A pre-segwit node with insufficiently validated blocks needs to redownload blocks"""
```
No need to have a run_test() method that only calls one method. Just put the logic in `run_test()`.

```suggestion
    def run_test(self):
        """A pre-segwit node with insufficiently validated blocks needs to redownload blocks"""
```

dhruv replied:
I am not sure I understand. It seems we call `main()` which then calls `test_framework.py::run_test()`. I am not sure how I could remove this `run_test()` and put the logic in `test_framework.py::run_test()`? 

Also, it seems `BitcoinTestFramework` [requires](https://github.com/bitcoin/bitcoin/blob/480bf01c295527bd212964efe4df3bb886db5654/test/functional/test_framework/test_framework.py#L69) overriding `run_test()`.

jnewbery replied:
Have you tried applying this diff?

dhruv replied:
You're right. My apologies for misreading the diff as just removing the `run_test()` function (which I did try manually). Updated. Thanks for your patience.

jnewbery commented on test/functional/feature_presegwit_node_upgrade.py:
```diff
@@ -0,0 +1,58 @@
+        # Node hasn't been used or connected yet
+        assert_equal(node.getblockcount(), 0)
+
+        self.restart_node(0, extra_args=["-segwitheight=10"])
```
Why not just start the node with `"-segwitheight=10"` first time, rather than starting and then restarting?

dhruv replied:
Nice! Done.

jnewbery commented:
Code review ACK 0dd10ccfd6ef13e5caa06fc380d6eb7f3a8fd22e

Looks great. Thanks @dhruv!

mzumsande commented on src/chainparams.cpp:
```diff
@@ -490,11 +490,8 @@ void CRegTestParams::UpdateActivationParametersFromArgs(const ArgsManager& args)
-            throw std::runtime_error(strprintf("Activation height %ld for segwit is out of valid range. Use -1 to disable segwit.", height));
-        } else if (height == -1) {
-            LogPrintf("Segwit disabled for testing\n");
-            height = std::numeric_limits<int>::max();
```
With the `segwitheight=-1` option removed, I think that also `IsScriptWitnessEnabled()` in `validation.cpp` has become obsolete and could be removed as well (the commentary there also seems to says that it's only kept along for `p2p_segwit.py`).

mzumsande replied:
There is another check for `consensusParams.SegwitHeight != std::numeric_limits<int>::max())` in `GenerateCoinbaseCommitment()`, also in validation.

dhruv replied:
Great catches. Thank you, @mzumsande! Both comments addressed.

mzumsande reviewed (commented):
Concept ACK.

dhruv committed:
[validation] Set witness script flag with p2sh for blocks

dhruv committed:
[validation] Always include merkle root in coinbase commitment

dhruv commented:
Addressed comments from @mzumsande. Ready for further review.

jnewbery commented:
utACK c44bac14b4

MarcoFalke commented on test/functional/p2p_segwit.py:
```diff
@@ -550,37 +535,9 @@ def test_v0_outputs_arent_spendable(self):
-                witness_root = CBlock.get_merkle_root([ser_uint256(0),
-                                                       ser_uint256(txid)])
-                script = get_witness_script(witness_root, 0)
-                assert_equal(witness_commitment, script.hex())
```
why is the test for default_witness_commitment removed?

jnewbery replied:
Segwit is a buried deployment. There are no miners calling `getblocktemplate` before the segwit activation is locked in.

MarcoFalke replied:
The commitment is produced regardless of whether segwit locked in or not.

jnewbery replied:
Oh, I see now that `default_witness_commitment` is not tested anywhere else. Perhaps it should be tested in mining_basic.py?

dhruv replied:
Thank you for the catch, @MarcoFalke ! Test is now moved into mining_basic.py

jnewbery commented:
@dhruv - do you plan to rebase this?

dhruv commented:
@jnewbery I do. Been on break for a long weekend where I live.

theStack commented:
Concept ACK

dhruv commented:
Rebased. Addressed comments. Ready for further review.

jnewbery commented on test/functional/mining_basic.py:
```diff
@@ -49,6 +51,9 @@ def set_test_params(self):
         self.setup_clean_chain = True
         self.supports_cli = False
 
+    def skip_test_if_missing_module(self):
+        self.skip_if_no_wallet()
```
This is ok for now, but it'd be nice to rework this test so that it doesn't require a wallet.

We could also update the test to verify that the witness commitment commits to the wtxids of the transactions in the block rather than the txids. Currently the block being tested contains only the coinbase transaction and one transaction that spends from a P2PKH output, so the txid is the same as wtxid. It could be updated to include transactions that spend segwit outputs so that the txid and wtxid are different.

dhruv replied:
I can do those in a follow-up PR. Can you tell me more about how I could remove the wallet dependency? My understanding is that to have `getblocktemplate` include a transaction that the witness root commits to, I need a valid transaction in the mempool, and to create that valid tx, I need a wallet. I feel like I am missing something about the test framework perhaps?

theStack replied:
@dhruv: Luckily we have a minimalistic wallet implementation for the functional test framework (first introduced by MarcoFalke last year, see #19800). Look for modules importing `MiniWallet` for example usage, see also issue #20078. Also, feel free to tag me in your follow-up PR, will be happy to review.

MarcoFalke replied:
Agree

jnewbery reviewed (commented):
utACK a806647d260132a00cd633160040625c7dd17803

Thanks for rebasing!

theStack reviewed (approved):
ACK a806647d260132a00cd633160040625c7dd17803

 Thanks to the [PR review club log](https://bitcoincore.reviews/21090) the changes were straight-forward to review. Built and ran tests locally, also checked via `git grep DeploymentEnabled.*SEGWIT` and `git grep GetLocalServices.*WITNESS` that all occurences of now-superfluous is-segwit-enabled checks were tackled.

mzumsande commented:
Code-Review ACK a806647d260132a00cd633160040625c7dd17803

laanwj commented:
Code review ACK a806647d260132a00cd633160040625c7dd17803, nice cleanup

MarcoFalke commented on test/functional/mining_basic.py:
```diff
@@ -89,7 +94,21 @@ def assert_submitblock(block, result_str_1, result_str_2=None):
+        witness_root = CBlock.get_merkle_root([ser_uint256(0),
+                                               ser_uint256(txid)])
+        script = get_witness_script(witness_root, 0)
+        assert_equal(witness_commitment, script.hex())
```
Python will throw key-error, so you don't need the "# Check that default_witness_commitment is present." section.

See:

```
>>> {'b':1}['a']
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 'a'
```

Suggestion:

```py
         assert_equal(tmpl['default_witness_commitment'], script.hex())
